### ISSUE TITLE
Lowercase roles and JWT tokens

### ISSUE THREAD

**Issue by JoshClose** [JudahGabriel/RavenDB.Identity#10]
I see in `UserStore.AddToRoleAsync`, the `roleName` is getting `ToLower` called on it and that is what is being stored in the users roles array.

I'm using `Microsoft.AspNetCore.Authentication.JwtBearer` for authentication. I create claims for the roles which are all lowercase. I have a role `SiteAdmin` that is stored as `siteadmin` under the user. The lowercase version is shipped off to the client. When the client makes requests, it sends that lowercase version to the server to authenticate.

I have controller actions that I'm decorating with `Authorize(Roles = "SiteAdmin)` because that is how I created it when doing `AddToRoleAsync(user, "SiteAdmin")`. This doesn't work though because the JWT token has the lowercase version in it.

Why are you doing a `ToLower` on the `roleName`? I did notice that it comes in normalized as an uppercase string, but curious if there is a specific reason you're lowering it. I also see that a role is created if it doesn't already exist, and that is using the lowercase name also. This seems a little odd especially if the user has extended `IdentityRole` and added some extra properties.

Since the `roleName` comes in normalized, do you see any issue with finding the role in the db first, and using the normal non-normalized name to store in `IdentityUser.Roles` instead so it matches `IdentityRole.Name`?

I believe all of this is only manifested because I'm using a JWT and it doesn't re-lookup the claims on an `Authorize`.

**Comment by JudahGabriel** [comment]
Hey Josh. Just got back from my travels, sorry for the late reply.

When I moved to MS Identity latest major version, I saw that the normalized name was being used through the framework. If I recall right, authorization wouldn't work if the role name differed by case.

I will re-validate that assumption today; if Identity framework doesn't care, I'll make it so that RavenDB.Identity doesn't care, and just uses whatever you pass in. 

I'll comment here later with the results.

**Comment by JoshClose** [comment]
I think there needs to be consistency between `User.Roles` and `Role.Name`. I could query the database and get the role names before adding it to the JWT, but it seems like they should match since it's sort of a foreign key.

I could be completely missing something too. These are just my observations so far.

**Comment by JudahGabriel** [comment]
Hi Josh,

Thanks for your patience on this one.

I just checked and confirmed that when we AddToRole is called, the role is _already normalized_ when I get it.

Here's how I verified that:

In my app's code, I called `userManager.AddToRoleAsync(user, "Manager"); // notice the casing`

That calls the AspNetCore Identity internal call, which forwards it on to our UserStore.AddToRoleAsync, passing in the role name:

![image](https://user-images.githubusercontent.com/312936/60139430-86cb1480-9773-11e9-9e48-6f5049b9eb1f.png)

Woops! Notice the roleName handed to us is now all caps.

This means AspNetCore Identity is internally converting your role to all caps _before_ I can even process it.

That's why I'm converting to lower-case; it's less ugly. 

Unfortunately, I think this means we can't store your original role name; we never see your original role name, only the .ToUpper() version of it handed to us by AspNetCore Identity framework internals.



This is really unfortunate

**Comment by JudahGabriel** [comment]
Ok, investigating this further, there might be one place where I see the not-yet-normalized role. Investigating...

**Comment by JudahGabriel** [comment]
OK, now that I've dug into this more and re-read your comment, I see the problem.

The problem is:

1. Create your own role (e.g. via RoleManager<IdentityUser>.CreateAsync(...)) with a case-varying role name (e.g. "SiteAdmin").
2. When you call userManager.AddToRoleAsync("SiteAdmin"), it stores the user, but uses the .ToLowered() role name (e.g. "siteadmin").
3. This breaks things relying on the case name, like JWT.

The fix for this is, when we call userManager.AddToRoleAsync(...), we should lookup the real name of the role ("SiteAdmin") and use that, rather than the normalized name.

**Comment by JudahGabriel** [comment]
I've fixed this issue and pushed out a new version, RavenDB.Identity 6.0.6, which includes the fix. 

If you want to see the fix in action, I've added an [MVC sample with authorization](https://github.com/JudahGabriel/RavenDB.Identity/tree/master/Samples/Mvc) that uses mixed-case role names (e.g. "Admin"). Running the sample will create the roles and users, and there will be consistency between role.Name and user.Roles; they'll both have preserved case of the role ("Admin").
